### PR TITLE
Add core classes and functions for online clustering diarizer part 2

### PR DESCRIPTION
**Pull request by tango4j**
Signed-off-by: Taejin Park <tango4j@gmail.com>

[[NeMo Online Speaker Diarization VIDEO DEMO LINK](
https://drive.google.com/file/d/15ro8HKmzevXh_wx9oTtfRUw9wNtj6LuJ/view?usp=share_link)]

# What does this PR do ?

Part 2 for adding core classes and functions for online clustering diarizer.
** As soon as  [PR 5526](https://github.com/NVIDIA/NeMo/pull/5526) gets merged, `Files changed` will be reduced to 2` **
Dividing the origin huge PR into two: 
(part 1: online clustering (nmesc_clustering.py) + unit tests: [PR 5526](https://github.com/NVIDIA/NeMo/pull/5526))
part 2: Segmentation + online diarizer class + unit tests

**Collection**: 

ASR

# Changelog 

the following will be included in this PR
#### `OnlineClusteringDiarizer` in online_diarizer.py (inherits ClusteringDiarizer) (will be included in part2)
online version of `ClusteringDiarizer` class in clustering_diarizer.py

#### `OnlineSegmentor` class in speaker_utils.py: (will be included in part2)
online version of speech segmentations

image

Checkout [Help Slides Link](https://docs.google.com/presentation/d/1LDUTIGRNVuLJsy81V3GaHZlFDUpxZ9rux-fNPfgcLmw/edit?usp=sharing) for better understanding on why online diarizer needs
(1) linear sum assignment (Hungarian alg)
(2) Interval merging
(3) merge_vector function
and data flow of class and functions.

# Usage

This is core classes and functions update.
Example script will be pushed in the second round of PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.
[Online Neural Diarization of Unlimited Numbers of Speakers](https://arxiv.org/pdf/2206.02432.pdf)

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/models/online_diarizer.py [hunk 1]
```diff
@@ -0,0 +1,573 @@
...
+        self.segment_range_ts = {self.uniq_id: {}}
+        self.segment_raw_audio = {self.uniq_id: {}}
+        self.segment_indexes = {self.uniq_id: {}}
+
+        for scale_idx, (window, shift) in self.multiscale_args_dict['scale_dict'].items():
+            self.multiscale_embeddings_and_timestamps[scale_idx] = [None, None]
+            self.emb_vectors[scale_idx] = None
+            self.time_stamps[scale_idx] = []
+            self.segment_range_ts[scale_idx] = []
+            self.segment_raw_audio[scale_idx] = []
```
Comment (nithinraok) [inline nemo/collections/asr/models/online_diarizer.py:200]: why is `segment_raw_audio` initialized with `uniq_id` but fetched by `scale_idx` key?

File: nemo/collections/asr/models/online_diarizer.py [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+        audio_signal_lens = torch.tensor([self.n_embed_seg_len for k in range(audio_signal.shape[0])]).to(self.device)
+        _, torch_embs = self._speaker_model.forward(input_signal=audio_signal, input_signal_length=audio_signal_lens)
+        return torch_embs
+
+    @timeit
+    def _extract_online_embeddings(
+        self, audio_signal: torch.Tensor, segment_ranges: torch.Tensor, embeddings
+    ) -> torch.Tensor:
+        """
+        Incrementally extract speaker embeddings based on audio_signal and segment_ranges varialbes.
```
Comment (nithinraok) [inline nemo/collections/asr/models/online_diarizer.py:401]: varialbes -> variables

File: nemo/collections/asr/models/online_diarizer.py [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+        Unlike offline speaker diarization, speaker embedding and subsegment ranges are not saved to disk.
+        Measures the mismatch between `segment_ranges` and `embeddings` then extract the necessary amount of
+        speaker embeddings.
+
+        Args:
+            audio_signal (Tensor):
+                Torch tensor containing timeseries audio signal
```
Comment (nithinraok) [inline nemo/collections/asr/models/online_diarizer.py:408]: timeseries -> time series

**Review by nithinraok — APPROVED** [review]
Minor comments. 
LGTM. Excellent work.

File: nemo/collections/asr/models/online_diarizer.py [hunk 4]
```diff
@@ -0,0 +1,573 @@
...
+import time
+from copy import deepcopy
+from typing import Dict
+
+import torch
+from omegaconf import DictConfig
+
+from nemo.collections.asr.models import ClusteringDiarizer
+from nemo.collections.asr.parts.utils.offline_clustering import get_scale_interpolated_embs, split_input_data
+from nemo.collections.asr.parts.utils.online_clustering import OnlineSpeakerClustering
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/models/online_diarizer.py:25]: Will change this later with other conventions, if needed.

File: nemo/collections/asr/models/online_diarizer.py (hunk 3 above)
Comment (tango4j) [reply nemo/collections/asr/models/online_diarizer.py:408]: fixed

File: nemo/collections/asr/models/online_diarizer.py (hunk 1 above)
Comment (tango4j) [reply nemo/collections/asr/models/online_diarizer.py:200]: Fixed this.

File: nemo/collections/asr/models/online_diarizer.py [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+        Initialize segment variables for each scale.
+        Note that we have `uniq_id` variable in case where multiple sessions are handled.
+        """
+        self.emb_vectors = {self.uniq_id: {}}
+        self.time_stamps = {self.uniq_id: {}}
+        self.segment_range_ts = {self.uniq_id: {}}
+        self.segment_raw_audio = {self.uniq_id: {}}
+        self.segment_indexes = {self.uniq_id: {}}
+
+        for scale_idx, (window, shift) in self.multiscale_args_dict['scale_dict'].items():
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/models/online_diarizer.py:195]: Fixed this

File: nemo/collections/asr/models/online_diarizer.py (hunk 2 above)
Comment (tango4j) [reply nemo/collections/asr/models/online_diarizer.py:401]: fixed this.

File: nemo/collections/asr/models/online_diarizer.py [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+    - The instance created from `OnlineClusteringDiarizer` sets aside a certain amount of memory
+      to provide the upcoming inference with history information
+
+    - There are two major modules involved: `OnlineSegmentor` and `OnlineSpeakerClustering`.
+        OnlineSegmentor: Take the VAD-timestamps and generate segments for each scale
+        OnlineSpeakerClustering: Update the entire speaker labels of the given online session
+                                 while updating the speaker labels of the streaming inputs.
+
+    - The overall diarization process is done by calling `diarize_step` function.
+      `diarize_step` function goes through the following stpes:
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/models/online_diarizer.py:75]: fixed this

File: nemo/collections/asr/models/online_diarizer.py [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+        Attributes:
+            use_temporal_label_major_vote (bool):
+                Boolean for whether to use temporal majority voting
+            temporal_label_major_vote_buffer_size (int):
+                buffer size for majority voting
+            base_scale_label_dict (dict):
+                Dictionary containing multiple speaker labels for major voting
+                Speaker labels from multiple steps are saved for each segment index.
+        """
+        self.temporal_label_major_vote_buffer_size = 11
```
[the remark this answers is not in the document]
Comment (tango4j) [reply nemo/collections/asr/models/online_diarizer.py:181]: changed this to get the config key from clustering.parameters

**Review by nithinraok — APPROVED** [review]

**Pull request merged**